03-client-api/04-new.md

```diff
@@ -0,0 +1,287 @@
+To understand what exactly this means, let's look at a detailed example.
+
+
+### Get Attributes by Value 
+Since all our clients are implemented similarly, the following piece of Python is representative:
+```
+# make sure you've run `pip3 install grakn-client` and have Grakn running
+from grakn.client import GraknClient, DataType
+client = GraknClient(uri="localhost:48555")
+with client.session(keyspace="test") as session:
+    with session.transaction().read() as tx:   
+        iter = **tx.get_attributes_by_value(“John”, DataType.STRING)**
+```
+
+Here, we want to retrieve all the attributes that have string values called “John” in the keyspace "test". 
+The first gRPC message created is a a `Transaction.Req` from `Session.proto`, which needs to have the `getAttributes_req` field populated.
+This, in turn has the type `GetAttributes.Req`, which has a single field called `value`. `value` is a `ValueObject`, 
+which is defined in the `Concept.proto` file (excerpt below):
+
+```
+message Concept {
+    string id = 1;
+    BASE_TYPE baseType = 2;
+
+    enum BASE_TYPE {
+        ...
+        ATTRIBUTE_TYPE = 3;
+        ...
+    }
+    ...
+}
+
+message ValueObject {
+    oneof value {
+        string string = 1;
+        bool boolean = 2;
+        int32 integer = 3;
+        int64 long = 4;
+        float float = 5;
+        double double = 6;
+        int64 date = 7; // time since epoch in milliseconds
+    }
+}
+```
+
+In this case, the `ValueObject` needs to the string field populated with “John”.
+
+In Python, printing the full constructed message should produce something like this:
+```
+{                            # type Transaction.Req
+  getAttributes_req {        # type GetAttributes.Req
+    value {                  # type ValueObject (from Concept.proto)
+      string : "John"
+    }
+  }
+}
+``` 
+
+gRPC implementations differ in how to compose these messages together, so check the relevant docs.
+ In python, each of these compound messages needs to be instantiated and embedded using `CopyFrom` or `MergeFrom` 
+([Python Protobuf docs](https://developers.google.com/protocol-buffers/docs/pythontutorial)).
+
+
+The message that is returned will be a `Transaction.Req`. But which field will be populated? 
+You can get this from our naming conventions: It should be the one with type `GetAttributes.Iter`, with a single field called `id`.
+
+```
+{                            # type Transaction.Res
+  getAttributes_iter {       # type GetAttributes.iter
+    id: 1
+  }
+}
+```
+
+### Lazy Responses
+One key reason for using bidirectional streaming is lazy evaluation of queries. The id returned above represents an iterator on the server,
+which we can repeatedly request to retrieve the actual `Attribute` instances. This can be wrapped up on the client side as a local iterator. 
+For instance, in Python, next element in an iterator by calling `next(attribute_iterator)`
+
+```
+...
```

Comment:
with a few additions to this code block, it can be made testable. also, looks like this code block isn't 1.5 compatible